crates/amaru-consensus/src/consensus/store.rs

```diff
@@ -37,8 +39,152 @@ impl Display for StoreError {
 /// A simple chain store interface that can store and retrieve headers indexed by their hash.
 pub trait ChainStore<H>: Send + Sync
 where
-    H: Header,
+    H: IsHeader,
 {
     fn load_header(&self, hash: &Hash<32>) -> Option<H>;
     fn store_header(&mut self, hash: &Hash<32>, header: &H) -> Result<(), StoreError>;
+
+    fn get_nonces(&self, header: &Hash<32>) -> Option<(Epoch, Nonces)>;
+    fn put_nonces(
+        &mut self,
+        header: &Hash<32>,
+        epoch: Epoch,
+        nonces: Nonces,
+    ) -> Result<(), StoreError>;
+}
+
+#[derive(Debug)]
+pub struct Nonces {
+    active: Nonce,
+    evolving: Nonce,
+    candidate: Nonce,
+    previous_epoch_last_header: Hash<32>,
+}
+
+impl<C> cbor::encode::Encode<C> for Nonces {
+    fn encode<W: cbor::encode::Write>(
+        &self,
+        e: &mut cbor::Encoder<W>,
+        ctx: &mut C,
+    ) -> Result<(), cbor::encode::Error<W::Error>> {
+        e.array(4)?;
+        e.encode_with(self.active, ctx)?;
+        e.encode_with(self.evolving, ctx)?;
+        e.encode_with(self.candidate, ctx)?;
+        e.encode_with(self.previous_epoch_last_header, ctx)?;
+        Ok(())
+    }
+}
+
+impl<'b, C> cbor::decode::Decode<'b, C> for Nonces {
+    fn decode(d: &mut cbor::Decoder<'b>, ctx: &mut C) -> Result<Self, cbor::decode::Error> {
+        d.array()?;
+        Ok(Nonces {
+            active: d.decode_with(ctx)?,
+            evolving: d.decode_with(ctx)?,
+            candidate: d.decode_with(ctx)?,
+            previous_epoch_last_header: d.decode_with(ctx)?,
+        })
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum NoncesError {
+    #[error("cannot find nonces: unknown parent {parent} from header {header}")]
+    UnknownParent { header: Hash<32>, parent: Hash<32> },
+
+    #[error("unknown header: {header}")]
+    UnknownHeader { header: Hash<32> },
+
+    #[error("no parent header for: {header} (where one is clearly expected)")]
+    NoParentHeader { header: Hash<32> },
+
+    #[error("{0}")]
+    StoreError(#[from] StoreError),
+}
+
+impl<H: IsHeader> Praos<H> for dyn ChainStore<H> {
+    type Error = NoncesError;
+
+    fn get_nonce(&self, header: &Hash<32>) -> Option<Nonce> {
+        self.get_nonces(header)
+            .map(|(_epoch, nonces)| nonces.active)
+    }
+
+    fn evolve_nonce(&mut self, header: &H) -> Result<(), Self::Error> {
+        let (epoch, is_within_stability_window) = nonce::randomness_stability_window(header);
+
+        let parent_hash = header.parent().ok_or_else(|| NoncesError::NoParentHeader {
+            header: header.hash(),
+        })?;
+
+        let (parent_epoch, parent) =
+            self.get_nonces(&parent_hash)
+                .ok_or_else(|| NoncesError::UnknownParent {
+                    header: header.hash(),
+                    parent: parent_hash,
+                })?;
+
+        // Compute the new evolving nonce by combining it with the current one and the header's VRF
```

Comment:
Shouldn't this comment be in the definition of `evolve` function?

Reply:
Most likely. The code was inline initially, and I later moved it into a function. The comment somehow stayed there. Will move.